### ISSUE TITLE
test flakes when `wait_for_service_endpoints_ready` helper function is used

### ISSUE THREAD

**Issue by ianvernon** [cilium/cilium#1442]
Build link: https://jenkins.cilium.io/job/cilium/job/cilium/job/master/587/execution/node/29/log/?consoleFull  

Failure:
```
 15:52:15 nc: can't connect to remote host (172.20.48.40): Connection refused
15:52:15 ------------------------------------------------------------------------
15:52:15                           K8s Test Failed
15:52:15 Unable to nc redis-master 6379
15:52:15 
15:52:15 ------------------------------------------------------------------------ 
```

Looks to be an issue with the helper functions, as the wait function passes almost instantaneously:
```
15:52:11 replicationcontroller "redis-master" created
15:52:11 service "redis-master" created
15:52:11 replicationcontroller "redis-slave" created
15:52:11 service "redis-slave" created
15:52:11 replicationcontroller "guestbook" created
15:52:11 service "guestbook" created
15:52:11 NAME                    READY     STATUS              RESTARTS   AGE       IP        NODE
15:52:11 app2-4193432878-nvdw7   0/1       Terminating         0          1m        <none>    k8s-1
15:52:11 guestbook-qbrzz         0/1       ContainerCreating   0          1s        <none>    k8s-2
15:52:11 redis-master-t8gdv      0/1       ContainerCreating   0          1s        <none>    k8s-1
15:52:11 redis-slave-335sw       0/1       ContainerCreating   0          1s        <none>    k8s-1
15:52:11 Waiting for guestbook pod to be Running...
15:52:12 Waiting for guestbook service endpoints to be ready
15:52:12 + wait_for_service_endpoints_ready default guestbook 3000
15:52:12 + set +x
15:52:13 + set +e
```

Add more verbose output to the helper functions to give updates about what's going on and fix them so that they don't prematurely exit.

**Comment by ianvernon** [comment]
I don't think this is an issue with the wait loops after all, actually. 
The output of the helper functions with `set -x` enabled in the scripts shows them behaving as they should in terms of what is being tested: https://jenkins.cilium.io/job/cilium/job/cilium/job/PR-1441/1/execution/node/29/log/?consoleFull

The logs from the failed build link that was filed with this ticket (https://jenkins.cilium.io/job/cilium/job/cilium/job/master/587/execution/node/29/log/?consoleFull) test that the port shows up in the list of Kubernetes services:

```
function wait_for_service_endpoints_ready {
  set +x
  check_num_params "$#" "3"
  local namespace="${1}"
  local name="${2}"
  local port="${3}"

  echo "Waiting for ${name} service endpoints to be ready"
  wait_specified_time_test "test \"\$(kubectl get endpoints -n ${namespace} ${name} | grep -c \":${port}\")\" -eq \"1\"" "5"
  set -x
}
```

Is this test enough? Should we be waiting for endpoints themselves to regenerate? The logs from the build show a lot of the endpoints being regenerated - should we wait for the endpoints to be regenerated here as well?

@aanm , @tgraf , what do you think?

**Comment by tgraf** [comment]
> Is this test enough? Should we be waiting for endpoints themselves to regenerate? The logs from the build show a lot of the endpoints being regenerated - should we wait for the endpoints to be regenerated here as well?

Good observation. I'm not actually sure what the wait is trying to guarantee. The endpoints appearing as a kubectl resource does not guarantee any application of it. The wait should really wait for the loadbalancer to show up in `cilium service list` and check whether both the frontend and required backends are there.

This will be a great check in general as it will also automatically ensure that we always configure the correct backends and not just one of them.

**Comment by ianvernon** [comment]
The current output of `cilium service list` is hard to parse. I've been working on basic checks in the branch for this ticket. I've created #1448 for adding output that is more easily parsable by our scripts. Once that is done, we can add more robust checks to our CI functions.
